woocommerce/woocommerce-ios
# [Mobile Payments] Prevent TTP reconnection if no location authorisation

Closes: #9773


## Description


We now reconnect to the reader automatically on foregrounding the app, if the device and store support it and TTP has been used before.

For users which grant `allow once` permissions, this means that in the next app session (i.e. after it's been force quit or kicked from memory), they’ll be asked to provide location permission without the context of a payment attempt, which is strange – the system permission alert pops up over whatever screen they happen to be on, interrupting their flow.

With this change, we only attempt to reconnect if we already have location permission – this means we no longer show a location permission request out of context.

Users who choose to `allow once` will still get the benefit of automatic reconnection when the app is foregrounded, but only after they first connect to the TTP reader in that app session. 

## Testing instructions


Using a US-based store with WCPay, an iPhone XS or newer, and iOS 16+

1. Launch the app
2. Take a tap to pay payment
3. Force close the app
4. Launch Settings and go to Woo's page
5. Change location permissions to `Ask Next Time Or When I Share`
6. Go back in to the app
7. Observe that you are **not** shown a Location permission alert
8. Navigate to `Menu > Payments > Collect Payment`
9. Collect a Tap to Pay on iPhone payment
10. Observe that you have to do a full reader connection (takes 7-12s) before the Apple payment screen is shown – i.e. the reader did not automatically reconnect
11. In this time you will be asked for location permissions – select `Allow once`
12. Background the app
13. Foreground the app, wait a few seconds, and take another TTP payment
14. Observe that you go directly to the Apple payment screen – i.e. the reader automatically reconnected when you foregrounded the app.

## Screenshots



---
